### DISCUSSION TITLE
Comments in energy_forcing (and subsequent)

### DISCUSSION THREAD

**Discussion by depion** [contrailcirrus/pycontrails#135] [Q&A]
**Description**

In the function `mean_energy_flux_per_m` it states:

`- Instead of taking an average of the previous and following segments, energy flux is only
      calculated for the following segment.`

I think this refers to to the multiplication with the segment length in `energy_forcing`, therefore I suggest moving the comment to this function. The same is also true for the other statement

`- Discontinuity is no longer set to 0 (this occurs directly in model :mod:`pycontrails.models.cocip.Cocip`)`

 which also makes more sense in the docstring of `energy_forcing`. 

Also, it might be worth stating in the docstring that the term "energy forcing" seems not to be consistently defined (e.g., in Schumann et al., "Atmospheric Physics: Background - Methods - Trends", 2012), energy forcing is defined as energy input per distance. Teoh et al. ("Targeted Use...", 2022) give different energy forcings (integrated, energy per flight distance, energy per contrail length). 

Finally, why is only the segment length at time 2 used for the calculation (`return energy_flux_per_m * seg_length_t2`), instead of a mean segment length?

**Comment by mlshapiro** [comment]
Thanks @depion for the notes!

Would you be willing to make the changes you suggest? I can give you write access to push up a branch with these changes, and then we can review in the PR and merge it in!

**Comment by mlshapiro** [comment]
We aim to define Energy Forcing in units of *Joules* as the total energy stored (or removed) by the contrail.

Energy forcing per unit distance is one of the primary output quantities - perhaps we term this "Unit energy forcing" or use the full term "Energy forcing per flight distance" every time we use it?

**Comment by depion** [comment]
> Would you be willing to make the changes you suggest?

Yes, I will try! (I need the right to folk the repository, right?)

> Energy forcing per unit distance is one of the primary output quantities

I don't understand. The energy forcing `ef` right now is the energy input of the full contrail length and not an energy forcing per distance, right? Do you want to include the energy forcing per distance as well?

> Finally, why is only the segment length at time 2 used for the calculation (return energy_flux_per_m * seg_length_t2), instead of a mean segment length?

Can you comment to that? It seems a bit inconsistent to me to average over contrail width and radiative forcing at both timesteps but not over the segment length...?

**Comment by zebengberg** [comment]
> Can you comment to that? It seems a bit inconsistent to me to average over contrail width and radiative forcing at both timesteps but not over the segment length...?

You're right that this is completely arbitrary. I'm not attached to the current conventions whatsoever (but I can't speak for everyone!). Many of these idiosyncrasies are vestiges from previous versions of the code. It would be better to be consistent here. It would be great if you can propose a change in a PR!

A few notes:

- As `dt_integration` gets small, the numeric difference between using `seg_length_t2` and `mean(seg_length_t1, seg_length_t2)` also grows small. In essence, I think we're discussing a "midpoint" vs "right-hand" Riemann sum here.
- If we could have a pycontrails convention for all of our Riemann sums, that'd be great. Something like "use a midpoint RS whenever possible".
- The default value for `dt_integration` is 30 minutes. In my opinion, this is somewhat too large, though Roger disagrees with me here.
- This situation may be more complicated if using a Runge Kutta integration scheme. So far, we've avoided this pycontrails.

**Comment by zebengberg** [comment]
> Yes, I will try! (I need the right to folk the repository, right?)

You don't need for fork ... you can just clone locally and make a new branch.

```sh
git clone ...
git checkout -b feature/ef-conventions
```

**Comment by zebengberg** [comment]
> I don't understand. The energy forcing `ef` right now is the energy input of the full contrail length and not an energy forcing per distance, right? Do you want to include the energy forcing per distance as well?

That's right --- `ef` associated to a waypoint is the total energy forcing of the segment associated with the waypoint. It's easy enough to convert this to a per distance unit. We don't do this explicitly in the `Cocip` model ... you can do it on your side with something like this:

```python
out = cocip.eval(source)
ef_per_m = out["ef"] / out["segment_length"]
```

I can create a better example if this doesn't make sense!

**Comment by depion** [comment]
> "use a midpoint RS whenever possible"

I guess of all simple possibilities this one would be the one which is least sensitive to the time step size. If one uses only the initial or the final conditions of the contrail segment, there could be an unnecessary large impact for sudden changes with time. As you said, differences all disappear in the limit of infinitesimal time steps, but that is not how CoCiP runs.

**Comment by zebengberg** [comment]
> As you said, differences all disappear in the limit of infinitesimal time steps, but that is not how CoCiP runs.

Just want to note here that you can adjust the `dt_integration` parameter in `Cocip`. I think our implementation runs fairly stably even for small values here (I think we've run with a 1 second`dt_integration` and didn't encounter any weird precision issues with float64s).

I support any changes you propose to make these numerical approximations more consistent / more precise.
